prog/weave-npc/main.go

```diff
@@ -312,16 +352,36 @@ func root(cmd *cobra.Command, args []string) {
 			}
 		},
 		UpdateFunc: func(old, new interface{}) {
+			defer func() {
+				if r := recover(); r != nil {
+					stopChan <- struct{}{}
+				}
+			}()
 			handleError(npc.UpdateNetworkPolicy(old, new))
 		},
 	}
 	npController = makeController(client.NetworkingV1().RESTClient(), "networkpolicies", &networkingv1.NetworkPolicy{}, npHandlers)
 
-	go nsController.Run(wait.NeverStop)
-	go podController.Run(wait.NeverStop)
-	go npController.Run(wait.NeverStop)
-
 	signals := make(chan os.Signal, 1)
+	stopChan = make(chan struct{})
+
+	go func() {
+		nsController.Run(stopChan)
+		signals <- syscall.SIGINT
+	}()
+	go func() {
+		podController.Run(stopChan)
+		signals <- syscall.SIGINT
+	}()
+	go func() {
+		npController.Run(stopChan)
+		signals <- syscall.SIGINT
+	}()
+	go func() {
+		<-stopChan
+		close(stopChan)
```

Comment:
Isn't it dangerous to close this while there are still goroutines that could send to it?

Reply:
well, the application is ending anyways, and I was considering waitgroups, but I'm not sure the harm here, as the application is completely shutting down at this point regardless.  I'm certainly not against adding a waitgroup, and waiting until all the others stop completely.

Reply:
Please don't.

Reply:
this has been removed.